### PR TITLE
Fix attention backends for models with per-layer head counts (num_attention_heads_per_layer)

### PR DESCRIPTION
**Pull request by Jiminator**
## Motivation

Both attention backends assume a uniform query head count across layers. Models that set `num_attention_heads_per_layer` (e.g. `poolside/Laguna-S-2.1`, 48–72 heads) break:

- **FlashInfer**: the prefill/decode indices updaters plan wrappers with the config-level `num_attention_heads` (48). `run()` silently accepts a 72-head q but only covers the planned work partition — output rows past ~qo·(48/72) of every prefill batch are corrupted on each >48-head layer (silent quality degradation).
- **Triton**: decode buffers (`attn_logits`, `attn_lse`) are sized with the config-level count while kernels index `layer.tp_q_head_num` → OOB writes → IMA / device asserts / CUBLAS failures.

## Modifications

Add `ModelConfig.get_max_num_attention_heads()` (max of `num_attention_heads_per_layer`, falling back to `num_attention_heads`) and use it at the four sizing/planning sites (FlashInfer prefill + decode updaters, Triton main + draft backends). Over-planning for smaller-head layers is safe: standalone `BatchPrefillWithPagedKVCacheWrapper` with plan=72/run=48 heads matches reference within bf16 noise.

## Accuracy Tests

`poolside/Laguna-S-2.1-NVFP4` (48–72 heads/layer), 1x RTX PRO 6000 (SM120), TP=1:

- Needle retrieval at 15k/40k/73k-token prompts: FlashInfer 1/3, 0/3, 2/3 → **3/3 everywhere**; Triton crash → **3/3 everywhere** (matches vLLM).
- Few-shot GSM8K 200q, FlashInfer: 0.33–0.35 accuracy / up to 0.245 invalid before → **0.64–0.71 / 0.000 invalid** after (Triton reference on the same stack: 0.79–0.83).
- No-regression check for uniform-head models: teacher-forced prompt logprobs (~15k tokens), stock vs patched, `google/gemma-4-12B-it` (Triton) and `Qwen/Qwen3-4B` (FlashInfer): **byte-exact, max |Δ| = 0**.

## Speed Tests and Profiling

N/A — for uniform-head models the computed values are unchanged; for per-layer-head models buffers/plans grow to the per-layer max (required for correctness).

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #30496951934](https://github.com/sgl-project/sglang/actions/runs/30496951934)
Latest PR Test (Extra): :x: [Run #30496951757](https://github.com/sgl-project/sglang/actions/runs/30496951757)

### CODE REVIEW

File: test/registered/unit/configs/test_model_config.py [hunk 1]
```diff
@@ -52,5 +53,25 @@ def test_does_not_misclassify_causal_gemma3(self):
         self.assertFalse(is_embedding_gemma(config))
 
 
+class TestGetMaxNumAttentionHeads(CustomTestCase):
```
Comment (Fridge003) [inline test/registered/unit/configs/test_model_config.py]: Nit, we don't need this subtest

File: python/sglang/srt/layers/attention/flashinfer_backend.py [hunk 2]
```diff
@@ -1481,7 +1481,8 @@ class FlashInferIndicesUpdaterDecode:
     def __init__(self, model_runner: ModelRunner, attn_backend: FlashInferAttnBackend):
         # Parse Constants
         self.num_qo_heads = (
-            model_runner.model_config.num_attention_heads // get_parallel().attn_tp_size
+            model_runner.model_config.get_max_num_attention_heads()
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/flashinfer_backend.py:1475]: Nit: Add one line of comment explaining why applying max_num_attention_heads here
